Changes to throwing

phil235 opened the pull request:
- Fixed catching banana creampie in mid throw (and similar objs). Fixes #3824 , Fixes #10536 , (hitting a wall now makes the smashed pie effect appear ON the wall)
- Some additional adjustments to toy/balloon.
- Moved last_bumped var from movable to living.
- Renamed /obj/machinery/field/proc/bump to bump_field() to avoid confusion with Bump(). Same thing with /obj/effect/mine/Bumped() -> triggermine().
- Fixes sprite when hunter leaps into a wall. Fixes #10428
- Removed some commented code here and there.
- Remove allow_spin var, throw_at now uses a spin argument for that. 
- Replaced the "throwpass" atom var by the pass flag LETPASSTHROW.
- A thrown mob hitting another mob now produces a sound; also thrown mob and target one no longer swap places even on help intent.
- A thrown mob now is lightly hurt (and weakened) if he hits a wall/mob/dense object. I Nerfed the damage when thrown mob hit wall.(20 -> 10)
- Thrown obj/mob no longer bounces off wall unless it's no grav. Heavy thrown items now push an unanchored obj/mob target.
- Fixes losing all momentum when getting out of a thrown closet. Fixes #6569
- A lot of work on throw_at, throw_impact, and hitby to make the code more OOP. Thrown items no longer collide with border items on the side. Fixes #10479. The default throw_at() now uses diagonal directions. 
- Fixes the killer crusher. Fixes #10507 
- Removes useless gateway/Crossed and gateway/Bumped().
- I added the item var "thrownby" to be able to continue to log the thrower of the item when it hits a mob. It removes the need for a thrower argument in throw_impact() and hitby().

phil235 committed:
Fixed catching banana creampie in mid throw (and similar objs). 3824 10536 (hitting a wall now makes the smashed pie effect appear ON the wall)
Moved last_bumped var from movable to living.
Renamed /obj/machinery/field/proc/bump to bump_field() to avoid confusion with Bump(). Same thing with /obj/effect/mine/Bumped() -> triggermine().
Fixes sprite when hunter leaps into a wall. 10428
Removed some commented code here and there.
Remove allow_spin var, throw_at now uses a spin argument for that.
Throwpass atom var is no longer used to check whether thrown stuff can pass stuff (now using CanPass() proc like everything else), the var is still used for some adjacency click check.
A thrown mob hitting another mob now produces a sound; also thrown mob and target one no longer swap places even on help intent.
A thrown mob now is lightly hurt (and weakened) if he hits a wall/mob/dense object. Nerf the damage when thrown mob hit wall.(20 -> 10)
Thrown obj/mob no longer bounces off wall unless it's no grav. Heavy thrown items now push an unanchored obj/mob target.
Fixes losing all momentum when getting out of a thrown closet. 6569
A lot of work on throw_at, throw_impact, and hitby to make the code more OOP. Thrown items no longer collide with border items on the side. 10479
Fixes the killer crusher. 10507

Iamgoofball commented on code/modules/power/singularity/singularity.dm:
```diff
@@ -26,7 +26,7 @@
 	var/last_failed_movement = 0//Will not move in the same dir if it couldnt before, will help with the getting stuck on fields thing
 	var/last_warning
 	var/consumedSupermatter = 0 //If the singularity has eaten a supermatter shard and can go to stage six
-	allow_spin = 0
```
hey hey hey

this creates lag

Iamgoofball commented on code/game/atoms_movable.dm:
```diff
@@ -154,57 +149,60 @@
-	src.throwing = 1
-	if(target.allow_spin) // turns out 1000+ spinning objects being thrown at the singularity creates lag - Iamgoofball
+	throwing = 1
+	if(spin) // turns out 1000+ spinning objects being thrown at the singularity creates lag - Iamgoofball
```
Where is this defined?

RemieRichards replied:
the proc args dummy.

RemieRichards commented:
@phil235 does the new throw_at() correctly use diagonal dirs? it was a key reason I wrote the horrors of the xenomorph override of throw_at(). if it does not I have to give that aspect a :-1: as it looks HORRIFIC without it.

phil235 commented:
Yeah, I fucked up the xeno throw_at(), I'll fix it.

MrPerson commented on code/game/atoms_movable.dm:
```diff
@@ -93,17 +93,13 @@
 		A.Bumped(src)
-	return
-	..()
-	return
```
rip in peace

On a more serious note, this shouldn't have the as mob|obj|turf|area bit.

phil235 replied:
I will remove it.

MrPerson commented on code/game/gamemodes/cult/cult_structures.dm:
```diff
@@ -46,11 +35,7 @@
 
 /obj/effect/gateway/Bumped(mob/M as mob|obj)
-	spawn(0)
-		return
```
why.jpg
Actually why is this proc even here? Seems superfluous at this point.

phil235 replied:
I thought this was here to override obj/Bumped() or effect/Bumped() or something. But apparently there's no non-empty parent Bumped(), so yeah let's remove.

MrPerson commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -82,12 +82,18 @@
 			M.client.perspective = MOB_PERSPECTIVE
+		if(throwing)
+			spawn(2)
+				step(M, dir)
```
But shouldn't the closet be open and thus non-dense at this point? How on earth would anything collide with it?

phil235 replied:
The mob comes out of the closet, makes one step forward, then the open closet (still thrown) moves forward and bumps into the mob. The mob is hit by it and the closet now moves backwards.

phil235 committed:
Fixes thrown closet opening to not use spawn().
Fixes throw_at() to look nicely for xeno leap by adding the diagonals_first argument.
Removes useless gateway/Crossed and gateway/Bumped().
Changes throw_at() to use diagonal directions.
I replaced the atom variable "throwpass" with the LETPASSTHROW pass_flags
I readded hit_check proc to catch things getting on the thrown thing's tile during its sleep(1) (especially needed for mobs running towards the thing), the only other possibility would've been to add throwing checks in atom/movable/Crossed() (called after every move) and I don't think it'd be worth it.
I added the item var "thrownby" to be able to continue to log the thrower of the item when it hits a mob. It removes the need for a thrower argument in throw_impact() and hitby().

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into CatchThrownPieFix

Conflicts:
	code/game/atoms_movable.dm
	code/game/objects/items/weapons/dice.dm
	code/game/objects/weapons.dm
	code/modules/flufftext/Hallucination.dm
	code/modules/mob/living/living_defense.dm
	code/modules/mob/living/living_defines.dm
	code/modules/reagents/reagent_containers.dm

phil235 committed:
I hate fixing merge conflicts.

phil235 committed:
A typo and a forgotten thing to do.

phil235 commented:
It should be ready for review.

bustygoku commented:
Conflict

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into CatchThrownPieFix

Conflicts:
	code/game/atoms.dm
	code/game/atoms_movable.dm

phil235 commented:
fixed
